### DISCUSSION TITLE
Format to save Recorded Track

### DISCUSSION THREAD

**Discussion by kirylkaveryn** [organicmaps/organicmaps#8939] [Ideas]
### Problem Description
To implement the `Track Recording` feature, it is necessary to determine the format in which the recorded track will be stored.
Currently, tracks are stored in KML format as an array of coordinates (according to the [documentation](https://developers.google.com/kml/documentation/kml_tut#paths) ) within the `<LineString><coordinates>` tag. However, this format does not allow for storing additional data such as timestamps and altitude, which are necessary for the full functionality of the feature.

### Competiors
Some applications store the recorder tracks in GPX format (OsmAnd), some in KML (MapsMe) and for some the internal storage is hidden (AllTrails, Strava, GuruMaps etc).

#### MapsMe allows to export track in KML:
[MM_Example.kml.zip](https://github.com/user-attachments/files/16581472/MM_Example.kml.zip)

```xml
  <Placemark>
    <name>MM track record sample 01</name>
    <Style><LineStyle>
      <color>FFC76E00</color>
      <width>5</width>
    </LineStyle></Style>
    <TimeStamp><when>2024-07-21T16:35:54.221Z</when></TimeStamp>
    <LineString><coordinates>44.796089,41.719319,419 44.796119,41.719355,419 44.796106,41.719304,419 44.796122,41.719372,420 
44.796132,41.719452,421 44.796218,41.719493,421 44.79629,41.719536,421 44.796344,41.719619,420 44.796234,41.719448,420 
...
44.796211,41.719364,420 44.796188,41.719287,420</coordinates></LineString>
    <ExtendedData xmlns:mwm="https://maps.me">
      <mwm:name>
        <mwm:lang code="default">MM track record sample 01</mwm:lang>
      </mwm:name>
      <mwm:description>
      </mwm:description>
      <mwm:localId>0</mwm:localId>
      <mwm:additionalStyle>
      </mwm:additionalStyle>
      <mwm:visibility>1</mwm:visibility>
      <mwm:trainInfo>
        <mwm:point>
          <mwm:timestamp>2024-07-21T16:28:03.999Z</mwm:timestamp>
        </mwm:point>
        <mwm:point>
          <mwm:timestamp>2024-07-21T16:28:19.999Z</mwm:timestamp>
        </mwm:point>
        <mwm:point>
          <mwm:timestamp>2024-07-21T16:28:33.999Z</mwm:timestamp>
        </mwm:point>
```

#### GuruMaps allow to export in multiple formats (KML, GPX, JSON):
[GuruMaps_Example.kml.zip](https://github.com/user-attachments/files/16581487/GuruMaps_Example.kml.zip)

```xml
<?xml version="1.0" encoding="utf-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:gom="https://gurumaps.app/kml/v1">
 <Document>
  <atom:author>
   <atom:name>Guru Maps Pro/5.5.4</atom:name>
  </atom:author>
  <Placemark>
   <name>Guru maps Track Record sample 01</name>
   <gx:TimeStamp>2024-07-21T16:27:28.441Z</gx:TimeStamp>
   <styleUrl>#TrackStyle_2196F3C8</styleUrl>
 <LineString>
  <tessellate>1</tessellate>
  <coordinates>44.7961038874,41.719389975,419.7067567872
44.7961038874,41.719389975,419.7067567872
44.7961229981,41.7192837762,420.0570680666
44.7961157897,41.7192831057,419.8267213869
...
44.7961813362,41.7195452497,421.7215578135
44.7961904725,41.7195407235,421.7028505381
44.7961987705,41.7195353172,421.8852541027

  </coordinates>
 </LineString>
 <gx:Track>
  <when>2024-07-21T16:27:25.389Z</when><gx:coord>44.7961038874 41.719389975 419.7067567872</gx:coord>
  <when>2024-07-21T16:27:31.393Z</when><gx:coord>44.7961038874 41.719389975 419.7067567872</gx:coord>
  <when>2024-07-21T16:27:33.657Z</when><gx:coord>44.7961229981 41.7192837762 420.0570680666</gx:coord>
...
  <when>2024-07-21T16:33:27.999Z</when><gx:coord>44.7961904725 41.7195407235 421.7028505381</gx:coord>
  <when>2024-07-21T16:33:28.999Z</when><gx:coord>44.7961987705 41.7195353172 421.8852541027</gx:coord>
 </gx:Track>
  </Placemark>
  <Style id="TrackStyle_2196F3C8">
   <LineStyle>
    <color>C8F39621</color>
    <width>5</width>
   </LineStyle>
  </Style>
 </Document>
</kml>

```

#### OsmAnd and Strava export only in GPX:
[Osmand_Example.gpx.zip](https://github.com/user-attachments/files/16581492/Osmand_Example.gpx.zip)
[Strava_Example.gpx.zip](https://github.com/user-attachments/files/16581500/Strava_Example.gpx.zip)

### Possible Solutions
1. It seems like the best format to store tracks is GPX because it supports waypoints, tracks, and paths https://www.topografix.com/GPX/1/1/#type_trkType.

2. OM stores the files in KML so it can be extended with the custom tags to support additional 
metatada so it will be possible to convert it in GPX during the exporting and vice versa.

3.

### Requirements (will be updated)
1. The recorded track KML should be stored with additional information (timestamps and altitudes)
2. OM should support tracks **with** and **without** extended metadata (in core)
3. OM should support track import from the different applications and formats (KML, GPS)
4. OM should support track export in proprietary KML (to share with the other OM users) and standard GPX

@biodranik @rtsisyk @vng @cyber-toad WDYT?

**Comment by biodranik** [comment]
@kirylkaveryn did you see https://github.com/organicmaps/organicmaps/issues/8314 ? We should leverage existing KML spec before inventing a wheel. Storing tracks internally in both GPX and KML formats is a more complex solution than using only KML. What are pros and cons?

**Reply by biodranik** [reply]
Just for reference, an example of a GPX track from the Motion-X app.
[Track_465 a.gpx.zip](https://github.com/user-attachments/files/16595812/Track_465.a.gpx.zip)
